File: backend/app/schemas/hyperparams.py
from typing import Literal, Optional

from app.schemas.common import CamelModel

ShapeKind = Literal["circle", "square", "hexagon", "octagon", "dodecagon"]

# L-BFGS-B is intentionally excluded -- see the comment at the top of
# app/core/packing.py for why it can't support this problem's constraints.
SolverMethod = Literal["slsqp", "cobyla", "trust-constr"]


class Hyperparams(CamelModel):
    n_restarts: int = 100
    n_refine: int = 10
    alpha: float = 100.0
    shape: ShapeKind = "octagon"
    # Hexagon-only: an extra 90-degree rotation on top of whatever the
    # symmetry mode already applies (see app/core/shapes.py's get_bases).
    hexagon_extra_rotation: bool = False
    # Square-only: rotates it 45 degrees into a diamond. A manual toggle,
    # not baked into get_bases -- any diagonal-symmetry default lives at the
    # frontend store's call site.
    square_extra_rotation: bool = False
    # Dodecagon-only: rotates it 15 degrees. A manual toggle, mirrors
    # square_extra_rotation's mechanism -- any diagonal-symmetry default
    # lives at the frontend store's call site.
    dodecagon_extra_rotation: bool = False
    seed: Optional[int] = None
    solver_method: SolverMethod = "slsqp"
    # Left unset (None) to use scipy's own per-method defaults.
    tol: Optional[float] = None
    max_iter: Optional[int] = None
    # Mirrors the frontend's rendering tolerances (see
    # frontend/src/types/hyperparams.ts) -- needed here too so
    # app/core/tiling_candidates.py's angular-coverage checks agree with the
    # active/semi-active split the user sees rendered live in the Packing
    # Editor (see tiling_service.py, which feeds these into
    # enumerate_tiling_candidates for the Tiling Editor's "Seed tiling" step).
    active_snap_length_tolerance: float = 0.1
    active_snap_angle_tolerance: float = 10.0
    # Only meaningful with initFrom=CURRENT (a re-optimize, not the first
    # solve): the largest per-restart random displacement applied to
    # position variables, ramping linearly from 0 (restart 0, the exact
    # current layout) up to this by the last restart -- see
    # solve_service.py's basin-hopping-style restart loop. Perturbation is
    # deliberately allowed to push a position outside [0,1].
    max_noise_amplitude: float = 0.2
    # Tiling solve (see app/core/tiling_solve.py): whether a flap's non-
    # paper-boundary-facing side must never have a gap of >=180 degrees
    # between selected crease directions (strict, the default -- required
    # for a genuinely concave/reflex vertex) or may relax to allowing a gap
    # of exactly 180 degrees. Turning this off can let the solver settle for
    # cheaper candidates on a flap that doesn't actually need full coverage.
    tiling_strict_concavity: bool = True

File: frontend/src/types/hyperparams.ts
import type { ShapeKind } from '../geometry/shapes'

/** L-BFGS-B is deliberately not an option — see the comment at the top of
 * backend/app/core/packing.py for why it can't support this problem. */
export type SolverMethod = 'slsqp' | 'cobyla' | 'trust-constr'

export interface HyperparamsState {
  nRestarts: number
  nRefine: number
  alpha: number
  shape: ShapeKind
  /** Hexagon-only: an extra 90° rotation on top of whatever the symmetry
   * mode already applies (see `geometry/shapes.ts`'s `hexagonBases`). */
  hexagonExtraRotation: boolean
  /** Square-only: rotates it 45° into a diamond (see `geometry/shapes.ts`'s
   * `squareBases`). Manual toggle, defaulted on (but still overridable) the
   * moment shape=square+symmetryMode=diagonal newly becomes true — see
   * `state/store.ts`'s `setHyperparams`/`setSymmetryMode`. */
  squareExtraRotation: boolean
  /** Dodecagon-only: rotates it 15° (see `geometry/shapes.ts`'s
   * `dodecagonBases`). Same manual-toggle-with-diagonal-default mechanism as
   * `squareExtraRotation`. */
  dodecagonExtraRotation: boolean
  seed?: number | null
  solverMethod: SolverMethod
  /** Left unset (null) to use scipy's own per-method defaults. */
  tol?: number | null
  maxIter?: number | null
  /** Rendering-only ("active snapping threshold") — never meaningfully sent
   * to the backend (extra hyperparams fields are ignored server-side): how
   * far a leaf pair's actual center distance may drift from the tree-implied
   * tangency distance (scale * tree distance), as a fraction of that
   * distance, and still be drawn as an active path in the packing canvas. */
  activeSnapLengthTolerance: number
  /** Rendering-only: for non-circle shapes, how far (in degrees) an active
   * path's angle may drift from the nearest shape-face-normal-perpendicular
   * multiple and still render as a solid ("fully active") line rather than a
   * dashed semi-active parallelogram. */
  activeSnapAngleTolerance: number
  /** Only meaningful for a re-optimize (not the very first solve): the
   * largest per-restart random displacement applied to position variables,
   * ramping linearly from 0 (restart 0, the exact current layout) up to
   * this by the last restart — see backend `solve_service.py`'s basin-
   * hopping-style restart loop. Perturbation is deliberately allowed to
   * push a flap outside [0,1]. */
  maxNoiseAmplitude: number
  /** Manual tiling editor: the minimum on-screen feature size (unit-square
   * units) below which the running cleanup pass (see `state/actions/
   * tilingGraphActions.ts`'s `runTilingCleanup`, called after every
   * pointerup in the tiling editor) snaps two near-coincident vertices
   * together or splits a leg a vertex has drifted onto. */
  tilingMinFeatureSize: number
  /** Backend tiling solve (see `backend/app/core/tiling_solve.py`): whether
   * a flap's non-paper-boundary-facing side must never have a gap of >=180
   * degrees between selected crease directions (strict, the default) or
   * may relax to allowing a gap of exactly 180 degrees. Sent wholesale to
   * the backend along with the rest of `HyperparamsState` -- see
   * `api/client.ts`. */
  tilingStrictConcavity: boolean
  /** Manual tiling editor: the cap on how many mirror-reflections a hinge
   * ray (see `geometry/hingeRayCast.ts`'s `castHingeRay`) will trace before
   * giving up and rendering whatever partial path it's traced so far --
   * raise this if a complex/dense tiling's hinge preview looks visibly
   * truncated. */
  tilingMaxHingeBounces: number
  /** Manual tiling editor: how close (in degrees) a candidate direct/bent
   * path may sit to an already-snapped direction (or exactly between two)
   * before it's hidden from the path-option picker as too visually
   * ambiguous to offer -- see `geometry/tilingGraphOps.ts`'s
   * `computePathOptions`. Mirrors `activeSnapAngleTolerance`'s role for the
   * Packing Editor's active-path rendering. */
  tilingPathOfferToleranceDeg: number
}

export const DEFAULT_HYPERPARAMS: HyperparamsState = {
  nRestarts: 20,
  nRefine: 1,
  alpha: 100,
  shape: 'circle',
  hexagonExtraRotation: false,
  squareExtraRotation: false,
  dodecagonExtraRotation: false,
  seed: null,
  solverMethod: 'slsqp',
  tol: null,
  maxIter: null,
  activeSnapLengthTolerance: 0.1,
  activeSnapAngleTolerance: 10,
  maxNoiseAmplitude: 0.2,
  tilingMinFeatureSize: 0.01,
  tilingStrictConcavity: true,
  tilingMaxHingeBounces: 60,
  tilingPathOfferToleranceDeg: 5,
}
